BootloaderCommonPkg/Library/SecureBootLib/SecureBootRsa.c

```diff
@@ -71,21 +71,36 @@ DoRsaVerify (
     return RETURN_INVALID_PARAMETER;
   }
 
-  Status =CalculateHash  (Data, Length, SignatureHdr->HashAlg, Digest);
-  if (EFI_ERROR(Status)) {
-    return RETURN_UNSUPPORTED;
-  }
-
-  if (OutHash != NULL) {
-    CopyMem (OutHash, Digest, DigestSize);
-  }
-
   DEBUG ((DEBUG_INFO, "SignType (0x%x) SignSize (0x%x)  SignHashAlg (0x%x)\n", \
                   SignatureHdr->SigType, SignatureHdr->SigSize, SignatureHdr->HashAlg));
 
   if(SignatureHdr->SigType == SIGNING_TYPE_RSA_PKCS_1_5) {
+    Status = CalculateHash  (Data, Length, SignatureHdr->HashAlg, Digest);
+    if (EFI_ERROR(Status)) {
+      return RETURN_UNSUPPORTED;
+    }
+
+    if (OutHash != NULL) {
+      CopyMem (OutHash, Digest, DigestSize);
+    }
+
     Status = RsaVerify_Pkcs_1_5 (PublicKey, SignatureHdr, Digest);
-  } else {
+
+  } else if(SignatureHdr->SigType == SIGNING_TYPE_RSA_PSS) {
+
+    // Calculate Hash only when OutHash is valid
+    // RSA PSS requires to pass message to be verified
+    if (OutHash != NULL) {
+      Status = CalculateHash  (Data, Length, SignatureHdr->HashAlg, Digest);

```

Comment:
Hashing will be done twice for RSA-PSS. Do we know where it will impact most?
What are the large components which will follow this path and will experience twice hashing?

Reply:
Hashing in case of PSS would be done for  CfgDatablob and KeyHashManifest verification would be twice.  Rest of the calls as Container verification for Boot image, hash calculation would  be done only only in IPP.

Reply:
CfgDataBlob and KeyHashMan sizes are pretty small...so double hashing may not have a impact.
For kernel image verification, won't that happen twice...as we extend the measurement of kernel image in TPM?

Reply:
For Kernel Image using container, 
- Container header is RSA verified. While extending to TPM, hash is recalculated. (size of header is small)
- Kernel image (component in container) is hash verified.  When extended to TPM it uses the hash in Container -Component header. In cases as SM3, hash would be calculated for all TPM measurements as its not supported while signing.

For IAS image method, yes there would double calc. But its getting deprecated,

Reply:
Sounds good.